lib/lf/fe/loc_comp_ellbvp.h

```diff
@@ -20,214 +20,236 @@
 #include "lagr_fe.h"
 
 namespace lf::fe {
-/** @brief Computing the element matrix for the (negative) Laplacian
- *         and linear finite elements.
- *
- * The main purpose of this class is to compute the element matrix for
- * the Laplacian on affine triangles or bilinearly mapped quadrilaterals.
- * These element matrices are provided by the `Eval()` method.
- *
- * @note the `Eval()` method will always return a _reference_ to a 4x4 matrix
- * also for triangles. In this case the last row and column must be ignored.
+
+/**
+ * @brief Preparing reusable information for parametric Lagrangian finite
+ * elements
  *
- * This class complies with the requirements for the type `ELEM_MAT_COMP`
- * given as a template parameter to define an incarnation of the function
- * AssembleMatrixLocally().
+ * For _uniform_ _parametric_ Lagrangian finite element spaces local
+ * computations applied to all celss of a mesh often rely on a single quadrature
+ * rule for a particular cell type. Therefore the values and gradients of
+ * reference shape functions can be precomputed at reference quadrature points
+ * and stored for reuse on every cell. This is the purpose of this class.
  */
-class LinearFELaplaceElementMatrix {
- public:
-  using ElemMat = Eigen::Matrix<double, 4, 4>;
+class LocCompLagrFEPreprocessor {
+ protected:
+  /** @brief Constructor: initializing local data
+   *
+   * @param fe_trie_p pointer to finite element to be used on triangles
+   * @param fe_quad_p pointer finite element for quadrilaterals
+   * @param loc_quad_order desired order of local quadrature, default value = 0.
+   *        If = 0, the quadrature order is determined from the polynomial
+   *        degree of the reference shape functions.
+   *
+   * @note null pointers mayb be passed. In this case the precomputations for
+   *       a particular type of element are skipped, an the number of local
+   * shape functions and quadrature points for that type is assumed to be zero.
+   *
+   * The two parametric finite elements for triangles and quadrilaterals have to
+   * be compatible in the sense that they all assign exactly one reference shape
+   * function to each vertex and the same number of interior shape functions
+   * to each edge.
+   *
+   * The constructor sets up local quadrature rules and fetches the values and
+   * gradients of references local shape functions in the quadrature points (on
+   * the reference element)
+   *
+   * The class also provides cell-type-controlled access methods to precomputed
+   * information on the reference element.
+   */
+  LocCompLagrFEPreprocessor(
+      std::shared_ptr<const ScalarReferenceFiniteElement<double>> fe_tria_p,
+      std::shared_ptr<const ScalarReferenceFiniteElement<double>> fe_quad_p,
+      lf::quad::quadOrder_t loc_quad_order = 0);
+
+  /** @brief type-dependent query of quadrature points */
+  inline Eigen::MatrixXd qr_points(lf::base::RefEl ref_el) {
```

Comment:
Instead of defining the methods `qr_points`, `qr_weights`, `qr_num_pts()` why not just define one function `qr()` which returns the quadrature rule for a given reference element?